src/main/java/org/spongepowered/api/text/Transformer.java

```diff
@@ -0,0 +1,292 @@
+/*
+ * This file is part of SpongeAPI, licensed under the MIT License (MIT).
+ *
+ * Copyright (c) SpongePowered <https://www.spongepowered.org>
+ * Copyright (c) contributors
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+package org.spongepowered.api.text;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+
+import java.util.Optional;
+import java.util.function.BiFunction;
+import java.util.function.Function;
+import java.util.function.Supplier;
+
+/**
+ * The context aware transformer interface represents a function that tries to
+ * convert the passed context data or select from them to dynamically calculate
+ * or generate content suiting the given context best. {@link Text.Placeholder}s
+ * use this interface to resolve their replacements for
+ * {@link Texts#format(Text, Function)}.
+ *
+ * @param <T> The type returned by this transformer
+ *
+ * @see Texts#format(Text, Function)
+ * @see Text.Placeholder
+ * @see Text.Placeholder#calculateReplacement(Function)
+ */
+public interface Transformer<T> {
+
+    /**
+     * Tries to transform the input context into appropriate content. Other
+     * transformations may take place after this. Implementations should be
+     * aware that sometimes their context data may not be available, in that
+     * case this method should return {@link Optional#empty()}.
+     *
+     * @param context The context function that contains the context data
+     * @return The non-null content for the given context or
+     *         {@link Optional#empty()}
+     *
+     * @see Text.Placeholder#calculateReplacement(Function)
+     */
+    Optional<T> transform(Function<String, ?> context);
```

Comment:
I don't think I'm convinced on the use case here of taking arbitrary functions as a context for transformation. This requires the implementation of a transformer to deal with the case that the function fails, and I don't think that's a good responsibility for it to have. How is this adding anything compared to using a `Map<String, ?>`?